File: two_sigma/financial_modeling_challenge/cpp/src/environment.cpp
#include "environment.h"
#include <fstream>
#include <sstream>
#include <iostream>
#include <algorithm>
#include <cmath>

namespace two_sigma {

Environment::Environment(const std::string& data_path) : data_path_(data_path), current_ts_idx_(0) {
    load_data();
}

void Environment::load_data() {
    std::ifstream file(data_path_);
    if (!file) {
        std::cerr << "Could not open " << data_path_ << std::endl;
        return;
    }

    std::string line, word;
    std::getline(file, line); // header

    std::vector<std::vector<double>> rows;
    while (std::getline(file, line)) {
        std::stringstream ss(line);
        std::vector<double> row;
        while (std::getline(ss, word, ',')) {
            row.push_back(std::stod(word));
        }
        rows.push_back(row);
    }

    size_t n_rows = rows.size();
    if (n_rows == 0) return;
    size_t n_cols = rows[0].size();

    all_features_.resize(n_rows, n_cols - 3);
    all_targets_.resize(n_rows);
    ids_.reserve(n_rows);
    timestamps_.reserve(n_rows);

    for (size_t i = 0; i < n_rows; ++i) {
        ids_.push_back(static_cast<int>(rows[i][0]));
        timestamps_.push_back(static_cast<int>(rows[i][1]));
        all_targets_(i) = rows[i][2];
        for (size_t j = 3; j < n_cols; ++j) {
            all_features_(i, j - 3) = rows[i][j];
        }
    }

    std::vector<int> unique_ts = timestamps_;
    std::sort(unique_ts.begin(), unique_ts.end());
    unique_ts.erase(std::unique(unique_ts.begin(), unique_ts.end()), unique_ts.end());
    unique_timestamps_ = unique_ts;

    split_ts_idx_ = unique_timestamps_.size() / 2;
    current_ts_idx_ = split_ts_idx_;
}

Observation Environment::reset() {
    current_ts_idx_ = split_ts_idx_;
    predictions_full_.clear();
    targets_full_.clear();

    Observation obs;
    
    // Training data: everything before split_ts_idx_
    int split_ts = unique_timestamps_[split_ts_idx_];
    std::vector<int> train_indices;
    for (size_t i = 0; i < timestamps_.size(); ++i) {
        if (timestamps_[i] < split_ts) train_indices.push_back(i);
    }

    obs.train_features.resize(train_indices.size(), all_features_.cols());
    obs.train_target.resize(train_indices.size());
    for (size_t i = 0; i < train_indices.size(); ++i) {
        obs.train_features.row(i) = all_features_.row(train_indices[i]);
        obs.train_target(i) = all_targets_(train_indices[i]);
    }

    // Current test data
    std::vector<int> test_indices;
    for (size_t i = 0; i < timestamps_.size(); ++i) {
        if (timestamps_[i] == split_ts) test_indices.push_back(i);
    }
    
    obs.test_features.resize(test_indices.size(), all_features_.cols());
    obs.ids.clear();
    for (size_t i = 0; i < test_indices.size(); ++i) {
        obs.test_features.row(i) = all_features_.row(test_indices[i]);
        obs.ids.push_back(ids_[test_indices[i]]);
        targets_full_.push_back(all_targets_(test_indices[i]));
    }

    return obs;
}

std::pair<Observation, double> Environment::step(const Eigen::VectorXd& predictions) {
    for (int i = 0; i < predictions.size(); ++i) {
        predictions_full_.push_back(predictions(i));
    }

    current_ts_idx_++;
    if (is_done()) return {Observation(), 0.0};

    int ts = unique_timestamps_[current_ts_idx_];
    Observation next_obs;
    std::vector<int> test_indices;
    for (size_t i = 0; i < timestamps_.size(); ++i) {
        if (timestamps_[i] == ts) test_indices.push_back(i);
    }

    next_obs.test_features.resize(test_indices.size(), all_features_.cols());
    for (size_t i = 0; i < test_indices.size(); ++i) {
        next_obs.test_features.row(i) = all_features_.row(test_indices[i]);
        next_obs.ids.push_back(ids_[test_indices[i]]);
        targets_full_.push_back(all_targets_(test_indices[i]));
    }

    return {next_obs, 0.0};
}

bool Environment::is_done() const {
    return current_ts_idx_ >= unique_timestamps_.size();
}

double Environment::calculate_final_score() const {
    if (predictions_full_.empty()) return 0.0;
    
    double mean_y = 0;
    for (double y : targets_full_) mean_y += y;
    mean_y /= targets_full_.size();

    double ss_res = 0, ss_tot = 0;
    for (size_t i = 0; i < predictions_full_.size(); ++i) {
        ss_res += std::pow(targets_full_[i] - predictions_full_[i], 2);
        ss_tot += std::pow(targets_full_[i] - mean_y, 2);
    }
    
    double r2 = 1.0 - (ss_res / ss_tot);
    return (r2 > 0 ? 1.0 : -1.0) * std::sqrt(std::abs(r2));
}

} // namespace two_sigma

File: two_sigma/financial_modeling_challenge/rust/src/environment.rs
use crate::Observation;
use csv::Reader;
use ndarray::{Array1, Array2};
use std::collections::HashSet;
use std::error::Error;

pub struct Environment {
    data_path: String,
    all_features: Array2<f64>,
    all_targets: Array1<f64>,
    timestamps: Vec<i32>,
    ids: Vec<i32>,
    unique_timestamps: Vec<i32>,
    current_ts_idx: usize,
    split_ts_idx: usize,
    predictions_full: Vec<f64>,
    targets_full: Vec<f64>,
}

impl Environment {
    pub fn new(data_path: &str) -> Result<Self, Box<dyn Error>> {
        let mut reader = Reader::from_path(data_path)?;
        let mut ids = Vec::new();
        let mut timestamps = Vec::new();
        let mut targets = Vec::new();
        let mut features_vec = Vec::new();
        let mut n_features = 0;

        for result in reader.records() {
            let record = result?;
            ids.push(record[0].parse()?);
            timestamps.push(record[1].parse()?);
            targets.push(record[2].parse()?);
            
            let row_features: Vec<f64> = record.iter().skip(3).map(|s| s.parse().unwrap()).collect();
            n_features = row_features.len();
            features_vec.extend(row_features);
        }

        let n_rows = ids.len();
        let all_features = Array2::from_shape_vec((n_rows, n_features), features_vec)?;
        let all_targets = Array1::from_vec(targets);

        let mut unique_ts_set: HashSet<i32> = timestamps.iter().cloned().collect();
        let mut unique_timestamps: Vec<i32> = unique_ts_set.into_iter().collect();
        unique_timestamps.sort();

        let split_ts_idx = unique_timestamps.len() / 2;

        Ok(Self {
            data_path: data_path.to_string(),
            all_features,
            all_targets,
            timestamps,
            ids,
            unique_timestamps,
            current_ts_idx: split_ts_idx,
            split_ts_idx,
            predictions_full: Vec::new(),
            targets_full: Vec::new(),
        })
    }

    pub fn reset(&mut self) -> Observation {
        self.current_ts_idx = self.split_ts_idx;
        self.predictions_full.clear();
        self.targets_full.clear();
        self.get_observation()
    }

    pub fn step(&mut self, predictions: Array1<f64>) -> (Option<Observation>, f64, bool) {
        for &p in predictions.iter() {
            self.predictions_full.push(p);
        }

        self.current_ts_idx += 1;
        if self.is_done() {
            return (None, 0.0, true);
        }

        let obs = self.get_observation();
        (Some(obs), 0.0, false)
    }

    pub fn is_done(&self) -> bool {
        self.current_ts_idx >= self.unique_timestamps.len()
    }

    fn get_observation(&mut self) -> Observation {
        let ts = self.unique_timestamps[self.current_ts_idx];
        
        // Find indices for training (all before split)
        let split_ts = self.unique_timestamps[self.split_ts_idx];
        let mut train_indices = Vec::new();
        for (idx, &t) in self.timestamps.iter().enumerate() {
            if t < split_ts {
                train_indices.push(idx);
            }
        }

        let mut train_feat_vec = Vec::new();
        let mut train_target_vec = Vec::new();
        for &idx in &train_indices {
            for col in 0..self.all_features.ncols() {
                train_feat_vec.push(self.all_features[[idx, col]]);
            }
            train_target_vec.push(self.all_targets[idx]);
        }
        let train_features = Array2::from_shape_vec((train_indices.len(), self.all_features.ncols()), train_feat_vec).unwrap();
        let train_target = Array1::from_vec(train_target_vec);

        // Current test data
        let mut test_indices = Vec::new();
        for (idx, &t) in self.timestamps.iter().enumerate() {
            if t == ts {
                test_indices.push(idx);
            }
        }

        let mut test_feat_vec = Vec::new();
        let mut test_ids = Vec::new();
        for &idx in &test_indices {
            for col in 0..self.all_features.ncols() {
                test_feat_vec.push(self.all_features[[idx, col]]);
            }
            test_ids.push(self.ids[idx]);
            self.targets_full.push(self.all_targets[idx]);
        }
        let test_features = Array2::from_shape_vec((test_indices.len(), self.all_features.ncols()), test_feat_vec).unwrap();

        Observation {
            train_features,
            train_target,
            test_features,
            ids: test_ids,
        }
    }

    pub fn calculate_final_score(&self) -> f64 {
        if self.predictions_full.is_empty() { return 0.0; }
        
        let n = self.targets_full.len();
        let mean_y = self.targets_full.iter().sum::<f64>() / n as f64;
        
        let mut ss_res = 0.0;
        let mut ss_tot = 0.0;
        for i in 0..self.predictions_full.len() {
            ss_res += (self.targets_full[i] - self.predictions_full[i]).powi(2);
            ss_tot += (self.targets_full[i] - mean_y).powi(2);
        }
        
        let r2 = 1.0 - (ss_res / ss_tot);
        r2.signum() * r2.abs().sqrt()
    }
}
